### PR TITLE
allowNetworks argument added in app action manifest[EXT-4509]

### PR DESCRIPTION
**Pull request by vikaskumr**
Adding new endpoint property `allowNetworks` in the action manifest file.

User should provide the list of allowed networks in action manifest file.

By default the below networks will always have access
`api.contentful.com, upload.contentful.com, 127.0.0.1`

### CODE REVIEW

File: packages/contentful--app-scripts/lib/utils.js [hunk 1]
```diff
@@ -13,6 +14,22 @@ const throwValidationException = (subject, message, details) => {
   throw new TypeError(message);
 };
 
+const validateIpAddress = (ipAddress) => {
```
Comment (kdamball) [inline packages/contentful--app-scripts/lib/utils.js:17]: can we add some unit tests for these? It's a big regex and brittle. Also the name can be `isValidIpAddress` to hint that we're returning boolean

File: packages/contentful--app-scripts/lib/utils.js [hunk 2]
```diff
@@ -81,7 +98,33 @@ function getActionsManifest() {
   ----------------------------`);
     console.log('');
 
-    return manifest.actions.map((action) => ({ parameters: [], ...action })); // adding required parameters
+    const actions = manifest.actions.map((action) => {
+      const allowedNetworks = Array.isArray(action.allowedNetworks)
+        ? [...action.allowedNetworks, 'api.contentful.com', 'upload.contentful.com', '127.0.0.1']
+        : ['api.contentful.com', 'upload.contentful.com', '127.0.0.1'];
+
+      const validAllowedNetworks = allowedNetworks
```
Comment (kdamball) [inline packages/contentful--app-scripts/lib/utils.js:106]: You can simplify this with `hasInvalidNetwork = allowedNetworks.map().some(netWork => !isValid())`

File: packages/contentful--app-scripts/lib/utils.js [hunk 3]
```diff
@@ -81,7 +98,33 @@ function getActionsManifest() {
   ----------------------------`);
     console.log('');
 
-    return manifest.actions.map((action) => ({ parameters: [], ...action })); // adding required parameters
+    const actions = manifest.actions.map((action) => {
+      const allowedNetworks = Array.isArray(action.allowedNetworks)
+        ? [...action.allowedNetworks, 'api.contentful.com', 'upload.contentful.com', '127.0.0.1']
```
[the remark this answers is not in the document]
Comment (vikaskumr) [reply packages/contentful--app-scripts/lib/utils.js:103]: ohh ok I thought we need to add the default values here :)

File: packages/contentful--app-scripts/lib/utils.js (hunk 1 above)
Comment (kdamball) [reply packages/contentful--app-scripts/lib/utils.js]: Dont forget IPv6

Comment (andipaetzold) [reply packages/contentful--app-scripts/lib/utils.js]: Isn't there a library for that? 😶 I wouldn't reinvent the wheel for such a big regex.

Comment (vikaskumr) [reply packages/contentful--app-scripts/lib/utils.js]: I will look for the library.

Comment (vikaskumr) [reply packages/contentful--app-scripts/lib/utils.js]: I searched for library, but unfortunately I couldn't find all the things which I need in one library

File: packages/contentful--app-scripts/lib/utils.js [hunk 4]
```diff
@@ -81,7 +108,31 @@ function getActionsManifest() {
...
+          )} Invalid IP address found in the allowedNetworks array for action "${action.name}".`
+        );
+        // eslint-disable-next-line no-process-exit
+        process.exit(1);
+      }
+
+      return {
+        parameters: [],
+        ...action,
+        allowedNetworks: allowedNetworks,
```
Comment (colomolo) [inline packages/contentful--app-scripts/lib/utils.js:131]:
```suggestion
        allowedNetworks,
```

File: packages/contentful--app-scripts/lib/utils.js [hunk 5]
```diff
@@ -13,6 +14,32 @@ const throwValidationException = (subject, message, details) => {
   throw new TypeError(message);
 };
 
+const isValidIpAddress = (address) => {
+  const addressRegex =
+    /^(?:(?:https?|ftp):\/\/)?(?:localhost|(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,6}|(?:\d{1,3}\.){3}\d{1,3}|(\[(?:[A-Fa-f0-9]{1,4}:){7}[A-Fa-f0-9]{1,4}\]|(?:[A-Fa-f0-9]{1,4}:){7}[A-Fa-f0-9]{1,4}))(?::\d{1,5})?$/;
```
Comment (ivo-toby) [inline packages/contentful--app-scripts/lib/utils.js:19]: looks like its matching for protocols, which you remove in ln 115, so why still match for those?

File: packages/contentful--app-scripts/lib/utils.js [hunk 6]
```diff
@@ -13,6 +14,32 @@ const throwValidationException = (subject, message, details) => {
   throw new TypeError(message);
 };
 
+const isValidIpAddress = (address) => {
```
Comment (ivo-toby) [inline packages/contentful--app-scripts/lib/utils.js:17]: Nit: technically this method name is incorrect, is also checks fqdn's, not only ip's

File: packages/contentful--app-scripts/lib/utills.test.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+const { isValidIpAddress, removeProtocolFromUrl } = require('./utils');
+
+describe('isValidIpAddress', () => {
+  it('returns true for a valid IP address', () => {
+    const result = isValidIpAddress('192.168.0.1');
+    assert.strictEqual(result, true);
+  });
+
+  it('returns true for a valid domain', () => {
+    const result = isValidIpAddress('http://google.com');
```
Comment (ivo-toby) [inline packages/contentful--app-scripts/lib/utills.test.js:12]: This should have failed, we don't want the protocol in there

File: packages/contentful--app-scripts/lib/utills.test.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    assert.strictEqual(result, true);
+  });
+
+  it('returns true for a valid ipv6 address', () => {
+    const result = isValidIpAddress('2001:0db8:85a3:0000:0000:8a2e:0370:7334');
+    assert.strictEqual(result, true);
+  });
+
+  it('returns true for a valid domain with port', () => {
+    const result = isValidIpAddress('http://google.com:4000');
```
Comment (ivo-toby) [inline packages/contentful--app-scripts/lib/utills.test.js:22]: Same

File: packages/contentful--app-scripts/lib/utills.test.js
Comment (ivo-toby) [inline packages/contentful--app-scripts/lib/utills.test.js:1]: Maybe worth testing for a valid ipv6 address with port, ie: `[2001:0db8:85a3:0000:0000:8a2e:0370:7334]:443`

**Review by ivo-toby — CHANGES_REQUESTED** [review]

File: packages/contentful--app-scripts/lib/utils.js (hunk 5 above)
Comment (vikaskumr) [reply packages/contentful--app-scripts/lib/utils.js:19]: Users can still have input the domains with protocol, according to them it is correct domain but for our case we don't need it so that is why I am removing protocol from this `removeProtocolFromUrl` function.

According to the ticket I 🤔 we can accept it but we can remove it internally and then I can send it to backend

File: packages/contentful--app-scripts/lib/utills.test.js (hunk 7 above)
Comment (vikaskumr) [reply packages/contentful--app-scripts/lib/utills.test.js:12]: I am removing removing protocol in `removeProtocolFromUrl` and since we are accepting the domain with protocol, it is correct unless it is otherwise.

@kdamball wdyt?

File: packages/contentful--app-scripts/lib/utils.js (hunk 5 above)
Comment (ivo-toby) [reply packages/contentful--app-scripts/lib/utils.js:19]: I see, but you are removing protocol with removeProtocolFromUrl before checking with this method right? 
It's not that big of a deal, it's just that it's checking for things that are not relevant.

File: packages/contentful--app-scripts/lib/utills.test.js (hunk 7 above)
Comment (ivo-toby) [reply packages/contentful--app-scripts/lib/utills.test.js:12]: Ok, I understand, but that implies things to be done in other methods that you are not testing here.

File: packages/contentful--app-scripts/lib/utils.js (hunk 5 above)
Comment (vikaskumr) [reply packages/contentful--app-scripts/lib/utils.js:19]: Yes, make sense

File: packages/contentful--app-scripts/package.json [hunk 9]
```diff
@@ -27,7 +27,7 @@
     "prettier": "prettier **/*.js --write --ignore-path .gitignore",
     "lint": "eslint ./lib",
     "lint:fix": "npm run lint -- --fix",
-    "test": "mocha ./lib/**/*.test.js ./lib/*.test.js ./utils/**/*.test.js",
```
Comment (kdamball) [inline packages/contentful--app-scripts/package.json:30]: Why is this deleted?

Comment (vikaskumr) [reply packages/contentful--app-scripts/package.json:30]: by mistake 😔

File: packages/contentful--app-scripts/lib/utils.js [hunk 10]
```diff
@@ -13,6 +13,25 @@ const throwValidationException = (subject, message, details) => {
...
+  const addressRegex =
+    /^(?:localhost|(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,6}|(?:\d{1,3}\.){3}\d{1,3}|(\[(?:[A-Fa-f0-9]{1,4}:){7}[A-Fa-f0-9]{1,4}\]|(?:[A-Fa-f0-9]{1,4}:){7}[A-Fa-f0-9]{1,4}))(?::\d{1,5})?$/;
+  return addressRegex.test(address);
+};
+
+const removeProtocolFromUrl = (url) => {
+  try {
+    const protocolRemovedUrl = url.replace(/^https?:\/\//, '');
+
+    const host = protocolRemovedUrl.split('/')[0];
```
Comment (colomolo) [inline packages/contentful--app-scripts/lib/utils.js:26]:
```suggestion
    return protocolRemovedUrl.split('/')[0];
```

File: packages/contentful--app-scripts/lib/utils.js [hunk 11]
```diff
@@ -13,6 +13,25 @@ const throwValidationException = (subject, message, details) => {
   throw new TypeError(message);
 };
 
+const isValidNetwork = (address) => {
+  const addressRegex =
+    /^(?:localhost|(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,6}|(?:\d{1,3}\.){3}\d{1,3}|(\[(?:[A-Fa-f0-9]{1,4}:){7}[A-Fa-f0-9]{1,4}\]|(?:[A-Fa-f0-9]{1,4}:){7}[A-Fa-f0-9]{1,4}))(?::\d{1,5})?$/;
+  return addressRegex.test(address);
+};
+
+const removeProtocolFromUrl = (url) => {
```
Comment (colomolo) [inline packages/contentful--app-scripts/lib/utils.js:22]: Maybe name it just `stripProtocol`?

File: packages/contentful--app-scripts/lib/utils.js [hunk 12]
```diff
@@ -81,7 +100,31 @@ function getActionsManifest() {
   ----------------------------`);
     console.log('');
 
-    return manifest.actions.map((action) => ({ parameters: [], ...action })); // adding required parameters
+    const actions = manifest.actions.map((action) => {
+      const allowedNetworks = Array.isArray(action.allowedNetworks) ? action.allowedNetworks : [];
+
+      const hasInvalidNetworks = allowedNetworks
+        .map(removeProtocolFromUrl)
+        .some((netWork) => !isValidNetwork(netWork));
```
Comment (colomolo) [inline packages/contentful--app-scripts/lib/utils.js:108]: I would replace `some` with `find` to be able to get exact invalid network and show it in error message below. Same computing price, but we can give a bit more info to the user that way.

File: packages/contentful--app-scripts/lib/utils.js [hunk 13]
```diff
@@ -81,7 +100,31 @@ function getActionsManifest() {
   ----------------------------`);
     console.log('');
 
-    return manifest.actions.map((action) => ({ parameters: [], ...action })); // adding required parameters
+    const actions = manifest.actions.map((action) => {
+      const allowedNetworks = Array.isArray(action.allowedNetworks) ? action.allowedNetworks : [];
```
Comment (colomolo) [inline packages/contentful--app-scripts/lib/utils.js:104]: Can map here instead of doing it twice below.
```suggestion
      const allowedNetworks = Array.isArray(action.allowedNetworks) ? action.allowedNetworks.map(removeProtocolFromUrl) : [];
```

File: packages/contentful--app-scripts/lib/utills.test.js [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    assert.strictEqual(result, 'example.com');
+  });
+
+  it('returns valid ipv6 address', () => {
+    const result = removeProtocolFromUrl('http://[2001:0db8:85a3:0000:0000:8a2e:0370]:7334');
+    assert.strictEqual(result, '[2001:0db8:85a3:0000:0000:8a2e:0370]:7334');
+  });
+
+  it('returns valid domain', () => {
+    const result = removeProtocolFromUrl('example.com');
```
Comment (colomolo) [inline packages/contentful--app-scripts/lib/utills.test.js:54]: Should have protocol here, right? Or we can test both cases – with and without a protocol.

File: packages/contentful--app-scripts/lib/utills.test.js [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    assert.strictEqual(result, 'example.com');
+  });
+
+  it('returns valid domain with port', () => {
+    const result = removeProtocolFromUrl('example.com:40');
```
Comment (colomolo) [inline packages/contentful--app-scripts/lib/utills.test.js:59]: Same here

File: packages/contentful--app-scripts/lib/utills.test.js (hunk 14 above)
Comment (colomolo) [reply packages/contentful--app-scripts/lib/utills.test.js:54]: Ah, nevermind, you test it above.

File: packages/contentful--app-scripts/lib/utils.js [hunk 16]
```diff
@@ -13,6 +13,25 @@ const throwValidationException = (subject, message, details) => {
...
 };
 
+const isValidNetwork = (address) => {
+  const addressRegex =
+    /^(?:localhost|(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,6}|(?:\d{1,3}\.){3}\d{1,3}|(\[(?:[A-Fa-f0-9]{1,4}:){7}[A-Fa-f0-9]{1,4}\]|(?:[A-Fa-f0-9]{1,4}:){7}[A-Fa-f0-9]{1,4}))(?::\d{1,5})?$/;
+  return addressRegex.test(address);
+};
+
+const removeProtocolFromUrl = (url) => {
+  try {
```
[the remark this answers is not in the document]
Comment (vikaskumr) [reply packages/contentful--app-scripts/lib/utils.js]: I was previously using URL library but after removing forgot to remove try/catch

**Review by kdamball — APPROVED** [review]
